requirements.txt

```diff
@@ -55,6 +55,7 @@ psycopg2==2.9.5 ; python_version == '3.11'
 psycopg2==2.9.9 ; python_version >= '3.12' # (Noble) Mostly to have a wheel package
 pyopenssl==21.0.0 ; python_version < '3.12'
 pyopenssl==24.1.0 ; python_version >= '3.12' # (Noble) min 23.2.0, pinned for compatibility with cryptography==42.0.8 and security patches
+pyotp==2.9.0
```

Comment:
do you really need `pyotp` library ? I think Odoo already use TOTP/HOTP for user login, wouldn't it be better to reuse their implementation ?
See: https://github.com/odoo/odoo/blob/master/addons/auth_totp/models/totp.py

Reply:
That would have been very practical, but they only implemented the method to check the validity of a provided code depending on the secret. 
They don't seem to have a method to generate one, as they apparently assume people will use Google Authenticator or some other tool to do so.

Reply:
In any case you should add a line in the debian/control (and fedora, ...) to add this new dependency
Also, the version of pyotp is 2.3.0 in ubuntu focal (python 3.10)(and 2.6.0 for python 3.11 but maybe useless to pin it) so you may 2.3.0 this version for python < 3.12 and 2.9.0 for >=3.12

Anyway we need to check that we want this new dependency, this needs to be validated by @antonylesuisse  in all cases

Reply:
Taking a quick look it wouldn't be too much lines to have an implem odoo side, looks like it could be done in 20-30 lines for the generation part

Reply:
> Taking a quick look it wouldn't be too much lines to have an implem odoo side, looks like it could be done in 20-30 lines for the generation part

seems like the best solution to me as well

Reply:
Also, I think that the currently implemented part odoo side could be use, so you may check with @xmo-odoo

Reply:
> @loouis-t I thought the point was to store an IoT secret on the Odoo database and share it with the IoT to then use it to generate the TOTP on odoo's side (or import it on Google authentificator) ?

We already share a scret with the IoT when we connect it (it is saved in the IoT `odoo.conf` as `token`). 
We the goal is to automate the generation of a TOTP on the DB side based on this secret, and check the correspondance on the IoT side.

Reply:
then this class should be enough no ? What are you missing ? If they share the same implementation and secret, the TOTP computation should be the same no ?

Reply:
This class only provide the TOTP verification part (`.match()`), not the generation one.
We need a method to generate the TOTP from the secret, to then be able to send it to the IoT.